src/kokkos/test/kokkos/VertexFinder_t.cc

```diff
@@ -0,0 +1,299 @@
+};
+
+struct ClusterGenerator {
+  explicit ClusterGenerator(float nvert, float ntrack)
+      : rgen(-13., 13), errgen(0.005, 0.025), clusGen(nvert), trackGen(ntrack), gauss(0., 1.), ptGen(1.) {}
+
+  void operator()(Event& ev) {
+    int nclus = clusGen(reng);
+    ev.zvert.resize(nclus);
+    ev.itrack.resize(nclus);
+    for (auto& z : ev.zvert) {
+      z = 3.5f * gauss(reng);
+    }
+
+    ev.ztrack.clear();
+    ev.eztrack.clear();
+    ev.ivert.clear();
+    for (int iv = 0; iv < nclus; ++iv) {
+      auto nt = trackGen(reng);
+      ev.itrack[nclus] = nt;
+      for (int it = 0; it < nt; ++it) {
+        auto err = errgen(reng);  // reality is not flat....
+        ev.ztrack.push_back(ev.zvert[iv] + err * gauss(reng));
+        ev.eztrack.push_back(err * err);
+        ev.ivert.push_back(iv);
+        ev.pttrack.push_back((iv == 5 ? 1.f : 0.5f) + ptGen(reng));
+        ev.pttrack.back() *= ev.pttrack.back();
+      }
+    }
+    // add noise
+    auto nt = 2 * trackGen(reng);
+    for (int it = 0; it < nt; ++it) {
+      auto err = 0.03f;
+      ev.ztrack.push_back(rgen(reng));
+      ev.eztrack.push_back(err * err);
+      ev.ivert.push_back(9999);
+      ev.pttrack.push_back(0.5f + ptGen(reng));
+      ev.pttrack.back() *= ev.pttrack.back();
+    }
+  }
+
+  std::mt19937 reng;
+  std::uniform_real_distribution<float> rgen;
+  std::uniform_real_distribution<float> errgen;
+  std::poisson_distribution<int> clusGen;
+  std::poisson_distribution<int> trackGen;
+  std::normal_distribution<float> gauss;
+  std::exponential_distribution<float> ptGen;
+};
+
+// a macro SORRY
+#define LOC_WS(M) ((char*)(ws_h.data()) + offsetof(WorkSpace, M))
+
+void test() {
+  Kokkos::View<ZVertices*, KokkosExecSpace> onGPU_d("onGPU_d", 1);
+  Kokkos::View<WorkSpace*, KokkosExecSpace> ws_d("ws_d", 1);
+
+  Event ev;
+
+  float eps = 0.1f;
+  std::array<float, 3> par{{eps, 0.01f, 9.0f}};
+  for (int nav = 30; nav < 80; nav += 20) {
+    ClusterGenerator gen(nav, 10);
+
+    for (int i = 8; i < 20; ++i) {
+      auto kk = i / 4;  // M param
+
+      gen(ev);
+
+      Kokkos::parallel_for(
+          "init", team_policy(1, 1), KOKKOS_LAMBDA(const member_type& team_member) {
+            onGPU_d(0).nvFinal = 0;
+            ws_d(0).ntrks = 0;
+            ws_d(0).nvIntermediate = 0;
+          });
+
+      std::cout << "v,t size " << ev.zvert.size() << ' ' << ev.ztrack.size() << std::endl;
+      auto nt = ev.ztrack.size();
+
+      auto ws_h = Kokkos::create_mirror_view(ws_d);
+      auto onGPU_h = Kokkos::create_mirror_view(onGPU_d);
+
+      // copy data to host mirror
+      ::memcpy(LOC_WS(ntrks), &nt, sizeof(uint32_t));
+      ::memcpy(LOC_WS(zt), ev.ztrack.data(), sizeof(float) * ev.ztrack.size());
+      ::memcpy(LOC_WS(ezt2), ev.eztrack.data(), sizeof(float) * ev.eztrack.size());
+      ::memcpy(LOC_WS(ptt2), ev.pttrack.data(), sizeof(float) * ev.eztrack.size());
+      // deep copy from host to device
+      Kokkos::deep_copy(ws_d, ws_h);
```

Comment:
Always passing the execution space to Kokkos operations allows to run the serial build of the unit test without GPU
```suggestion
      Kokkos::deep_copy(KokkosExecSpace(), ws_d, ws_h);
```

Reply:
I'm a bit confused: we've passed execution space when declaring `ws_d` so we still need to explicitly indicate the execution space here?

Reply:
Strictly speaking `ws_d` definition has a memory space (when passed an execution space, the default memory space of that execution space is used). The purpose of the execution space object here is different. It makes `deep_copy` to `fence()` only that execution space (which allows to run a serial program on a machine without GPU even when the Kokkos runtime was built with CUDA enabled). My understanding is that when a CUDA execution space constructed with a CUDA stream is used, the `deep_copy()` is essentially `cudaMemcpyAsync()` without fencing.

Reply:
Thanks for the explanation. Done.